src/datasets/arrow_writer.py

```diff
@@ -428,18 +428,12 @@ def write_examples_on_file(self):
             return
 
         # order the columns properly
-        if self.schema:
-            schema_cols = set(self.schema.names)
-            common_cols, extra_cols = [], []
-            for col in self.current_examples[0][0]:
-                if col in schema_cols:
-                    common_cols.append(col)
-                else:
-                    extra_cols.append(col)
-            cols = common_cols + extra_cols
-        else:
-            cols = list(self.current_examples[0][0])
-
+        cols = (
+            [col for col in self.schema.names if col in self.current_examples[0][0]]
+            + [col for col in self.current_examples[0][0].keys() if col not in self.schema.names]
+            if self.schema
+            else self.current_examples[0][0].keys()
+        )
```

Comment:
```suggestion
        if self.schema:
            schema_cols = set(self.schema.names)
            examples_cols = self.current_examples[0][0].keys()
            common_cols = [col for col in self.schema.names if col in examples_cols]
            extra_cols = [col for col in examples_cols if col not in schema_cols]
            cols = common_cols + extra_cols
        else:
            cols = list(self.current_examples[0][0])
```

Reply:
What about using a set also for `examples_cols`?

Why cast `cols` to a list? And wouldn't it be clearer if we used `.keys()` there as well?

(if you agree with any change, please send a new code snippet suggestion so when done I can commit it)

Reply:
Unlike `set()`, `.keys` guarantees the same order of the elements as in the input dictionary.

Reply:
> Why cast cols to a list? And wouldn't it be clearer if we used .keys() there as well?

To ensure `cols` is a list in both scenarios (this step is not required, but is usually a good practice to avoid type-checking errors, for instance)

Reply:
> Unlike set(), .keys guarantees the same order of the elements as in the input dictionary.

What if you had 2 copies of it? One for checking (`col in examples_cols`), and another for iterating it.

Reply:
> To ensure `cols` is a list in both scenarios (this step is not required, but is usually a good practice to avoid type-checking errors, for instance)

I feel like it's unnecessary because it's gonna do an extra memory allocation (which also takes time, depending on the size) and extra iteration, but in the end, `cols` will be iterated only once later.